app/components/shared/process_steps_component.rb

```diff
@@ -6,13 +6,27 @@ def initialize(process:, service:, title:)
   end
 
   def render?
-    @process.blank? || %w[create review].include?(@process.state)
+    @process.blank? || %w[create draft review].include?(process_state)
   end
 
   def current_step_number
     @service.current_step_number
   end
 
+  def completed_step_number
+    if @process.is_a?(JobApplication)
+      @process.completed_steps.map { |step| @service.steps[step.to_sym][:number] }.max
+    else
+      @process.completed_step
+    end
+  end
+
+  def process_state
+    return @process.status if @process.is_a?(JobApplication)
```

Comment:
This component used to be tied too closely to Vacancy, so that it only worked when used with Vacancies.

With the changes in this file, it becomes tied to both Vacancy and JobApplication. This is a step in the wrong direction: it's become less abstract.

Ideally, it should know almost nothing about the implementation of the things it's consuming. Imagine that we are planning to share this ProcessSteps component across government so that any website can use it, without requiring Vacancy or JobApplication - because that is in fact what we hope to do to our best components!

Ideally, the component would be 'abstract' in that you can plug anything in (a vacancy, a job application, a petition form, an ice cream order form) and, as long as the thing you plug in is plug-shaped (i.e. is able to answer to a few particular methods specified by the component), the component can render the steps.

For example, we could say that anything that gets plugged in to the socket of ProcessStepsComponent should be able to answer to just one out of `status` or `state`. An ice cream order form could have a state, and then we'd be able to plug that in, without writing a series of lines like

```ruby
return @process.stage if @process.is_a?(IceCream)
return @process.point_we_got_up_to if @process.is_a?(PersonalProtectiveEquipmentOrder)
return @process.step if @process.is_a?(HowToPaintCourse)
```

Probably we want to change JobApplication's `status` column to be called `state` to match Vacancy (Vacancy already has one of each).

Reply:
Another example: you could make ProcessStepsComponent more abstract by moving this logic into a method that lives on JobApplication. Then ProcessStepsComponent just needs to call that method on the `@process`.

```ruby
if @process.is_a?(JobApplication)
   @process.completed_steps.map { |step| @service.steps[step.to_sym][:number] }.max
```

Reply:
I'm was happy to let this fly at the minute since I was going to refactor it soon anyway - so I will approve but feel free to request changes if you want this refactoring to happen right away!

Reply:
I don't mind who does it, but perhaps it would be an interesting job for @josephhull676?

Reply:
Maybe we can pair on it @josephhull676?